File: Programmers/Lv3/순위.js
function solution(n, results) {
  var answer = 0;
  var winGraph = [];
  var loseGraph = [];

  results.forEach((res) => {
    let winner = res[0];
    let loser = res[1];
    if (winGraph[winner]) {
      winGraph[winner].push(loser);
    } else {
      winGraph[winner] = [loser];
    }

    if (loseGraph[loser]) {
      loseGraph[loser].push(winner);
    } else {
      loseGraph[loser] = [winner];
    }
  });

  const bfs = (graph, start) => {
    var visited = new Set();
    var q = [start];
    visited.add(start);

    let res = 0;

    while (q.length > 0) {
      const node = q.shift();
      for (let nxtNode of graph[node] || []) {
        if (!visited.has(nxtNode)) {
          visited.add(nxtNode);
          q.push(nxtNode);
          res++;
        }
      }
    }
    return res;
  };

  for (let v = 1; v <= n; v++) {
    if (bfs(winGraph, v) + bfs(loseGraph, v) === n - 1) {
      answer++;
    }
  }

  return answer;
}

File: Programmers/Lv3/순위.py
from collections import deque

def bfs(graph, s):
    visited = set()
    q = deque()
    q.append(s)
    
    cnt = 0
    visited.add(s)
    
    while q:
        v = q.popleft()
        for adj in graph[v]:
            if adj not in visited:
                visited.add(adj)
                q.append(adj)
                cnt+=1
    return cnt

def solution(n, results):
    lose = [[] for _ in range(n + 1)]
    win = [[] for _ in range(n + 1)]
    
    for a, b in results:
        win[a].append(b)
        lose[b].append(a)
    
    res = 0
    for node in range(1, n + 1):
        if (bfs(win, node) + bfs(lose, node)) == n - 1:
            res+=1
        
    return res